switchio/connection.py

```diff
@@ -52,12 +54,79 @@ def run_in_order_threadsafe(awaitables, loop, timeout=0.5, block=True):
     return future
 
 
-class AsyncIOConnection(object):
+async def try_connect(host, port, password, prot, loop, log):
+    """Try to create a connection and authenticate to the
+    target FS ESL.
+    """
+    msg = ("Failed to connect to server at '{}:{}'\n"
+           "Please check that FreeSWITCH is running and "
+           "accepting ESL connections.".format(host, port))
+    for _ in range(5):
+        try:
+
+            await asyncio.wait_for(
+                loop.create_connection(lambda: prot, host, port),
+                timeout=0.1)
+            break
+        except (
+            ConnectionRefusedError, asyncio.TimeoutError, OSError,
+        ) as err:
+            log.warning(
+                "Connection to {}:{} failed, retrying..."
+                .format(host, port)
+            )
+    else:
+        raise ConnectionError(msg.format(host, port))
+
+    # TODO: consider using the asyncio_timeout lib here
+    try:
+        await asyncio.wait_for(prot.authenticate(), 10)
+    except asyncio.TimeoutError:
+        raise ConnectionRefusedError(msg.format(host, port))
+
+
+async def async_reconnect(host, port, password, prot, loop, log):
+    log.info("Attempting to reconnect to {}:{}".format(host, port))
+    if not prot.autorecon:
+        log.debug("Autorecon had been disabled")
+        return
+    else:
+        count = prot.autorecon
+
+    for i in range(count):
+        try:
+            await try_connect(host, port, password, prot, loop, log)
```

Comment:
Loop inside a loop?

Honestly, might be easier to just wrap this whole thing in an `async_timeout`, use an absolute timeout instead of faking it with iterations, and just do a while loop:

```python
with async_timeout(10):
    while True:
        await try_connect(...)
        break
```

Error/timeout handling missing, obviously

Reply:
Yes, `try_reconnect` is a longer running task that attempts to call the normal connection logic asynchronously when the connection drops.

Which part are you saying to wrap? `try_connect` just tries 5 times before baling whereas `try_reconnect` is waiting up to `autorecon` seconds.